### PR TITLE
Implement parameterized structs

### LINKED ISSUE #255: Parameterised struct formats
It would be nice to have parameters on struct formats. For example:

```
struct MyTable (SubTable : Format) : Format {
    count : U32Be,
    sub_tables : FormatArray count SubTable,
}
```

The parameters should form a _telescope_, allowing for dependencies between parameters:

```
struct Data (len : Int) (Data : FormatArray len U8) : Format {
    data : Data,
}
```

### PR DESCRIPTION
**Pull request by brendanzab**
This PR implements parameterized struct types and struct formats. This allows us to finally start implementing part of OpenType, which is extremely exciting! 

An example of how this looks in practice is:

```rust
struct GenericPoint (Coordinate : Format) : Format {
    x : Coordinate,
    y : Coordinate,
}

struct MyGenericArray (len : Int) (Coordinate : Format) : Format {
    data : FormatArray len (GenericPoint Coordinate),
}
```

As part of this change we also allow the representations of struct formats to be introduced and eliminated. This had been previously documented, but we had somehow missed implementing it.

Closes #255.

### CODE REVIEW

**Comment by brendanzab** [comment]
I still need to check the OpenType example in CI, so we should hold off merging until then!

File: fathom/src/lang/core.rs [hunk 1]
```diff
@@ -83,6 +83,8 @@ pub struct StructType {
     pub doc: Arc<[String]>,
     /// Name of this definition.
     pub name: String,
+    /// Parameter telescope.
```
Comment (brendanzab) [inline fathom/src/lang/core.rs:86]: Ahh, I guess this isn't really a telescope just yet - the parameters can't depend on each other yet!

File: fathom/src/lang/core.rs [hunk 2]
```diff
@@ -83,6 +83,8 @@ pub struct StructType {
...
+    pub params: Arc<[(Located<String>, Arc<Term>)]>,
```
Comment (brendanzab) [inline fathom/src/lang/core.rs:87]: Ahh, I guess this isn't actually a telescope just yet - the parameters can't depend on each other yet! We'd need to implement dependent function types first (see #281).

File: fathom/src/lang/core/typing.rs [hunk 3]
```diff
@@ -331,25 +355,22 @@ impl<'me> Context<'me> {
...
-                            let r#type =
-                                self.eval_with_locals(&mut type_locals, &field_declaration.type_);
-                            // TODO: stop on errors
-                            self.check_type(&field_definition.term, &r#type);
-                            let value = self.eval(&field_definition.term);
 
-                            type_locals.push(value);
+                field_declarations.for_each_field(
+                    self.globals,
+                    &self.item_definitions.clone(), // FIXME: avoid clone
```
Comment (brendanzab) [inline fathom/src/lang/core/typing.rs:361]: It's kind of frustrating to have to do these clones of the item definitions - but it's required seeing as we take `self` by mutable reference in the closure, even though we don't mutate `items` at all. Perhaps we could use [`elsa::FrozenMap`](https://docs.rs/elsa/1.4.0/elsa/map/struct.FrozenMap.html) to make this a bit nicer?

**Review by mikeday — APPROVED** [review]
Looks good!

**Pull request merged**